source/distributed/error_predictor.cc

```diff
@@ -0,0 +1,363 @@
+      register_data_attach();
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    ErrorPredictor<dim, spacedim>::register_data_attach()
+    {
+      // TODO: casting away constness is bad
+      parallel::distributed::Triangulation<dim, spacedim> *tria =
+        (dynamic_cast<parallel::distributed::Triangulation<dim, spacedim> *>(
+          const_cast<dealii::Triangulation<dim, spacedim> *>(
+            &dof_handler->get_triangulation())));
+      Assert(tria != nullptr, ExcInternalError());
+
+      handle = tria->register_data_attach(
+        [this](
+          const typename Triangulation<dim, spacedim>::cell_iterator &cell_,
+          const typename Triangulation<dim, spacedim>::CellStatus     status) {
+          return this->pack_callback(cell_, status);
+        },
+        /*returns_variable_size_data=*/false);
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    ErrorPredictor<dim, spacedim>::prepare_for_coarsening_and_refinement(
+      const Vector<float> &in,
+      const double         gamma_p,
+      const double         gamma_h,
+      const double         gamma_n)
+    {
+      std::vector<const Vector<float> *> all_in(1, &in);
+      prepare_for_coarsening_and_refinement(all_in, gamma_p, gamma_h, gamma_n);
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    ErrorPredictor<dim, spacedim>::unpack(std::vector<Vector<float> *> &all_out)
+    {
+      Assert(error_indicators.size() == all_out.size(),
+             ExcDimensionMismatch(error_indicators.size(), all_out.size()));
+
+      // TODO: casting away constness is bad
+      parallel::distributed::Triangulation<dim, spacedim> *tria =
+        (dynamic_cast<parallel::distributed::Triangulation<dim, spacedim> *>(
+          const_cast<dealii::Triangulation<dim, spacedim> *>(
+            &dof_handler->get_triangulation())));
+      Assert(tria != nullptr, ExcInternalError());
+
+      tria->notify_ready_to_unpack(
+        handle,
+        [this, &all_out](
+          const typename Triangulation<dim, spacedim>::cell_iterator &cell_,
+          const typename Triangulation<dim, spacedim>::CellStatus     status,
+          const boost::iterator_range<std::vector<char>::const_iterator>
+            &data_range) {
+          this->unpack_callback(cell_, status, data_range, all_out);
+        });
+
+      for (typename std::vector<Vector<float> *>::iterator it = all_out.begin();
+           it != all_out.end();
+           ++it)
+        (*it)->compress(::dealii::VectorOperation::insert);
+
+      error_indicators.clear();
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    ErrorPredictor<dim, spacedim>::unpack(Vector<float> &out)
+    {
+      std::vector<Vector<float> *> all_out(1, &out);
+      unpack(all_out);
+    }
+
+
+
+    template <int dim, int spacedim>
+    std::vector<char>
+    ErrorPredictor<dim, spacedim>::pack_callback(
+      const typename Triangulation<dim, spacedim>::cell_iterator &cell_,
```

Comment:
same here:
```suggestion
      const typename Triangulation<dim, spacedim>::cell_iterator &cell,
```

Reply:
We convert the `Triangulation::cell_iterator` to a `DoFHandler::cell_iterator` below, which will be assigned to that name. I would leave it like this here.